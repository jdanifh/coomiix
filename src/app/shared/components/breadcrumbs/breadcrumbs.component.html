  @for(breadcrumb of breadcrumbs(); track breadcrumb.title; let last = $last) {
    @if(!last) {
      <h3><a [routerLink]="[breadcrumb.path]" routerLinkActive="active" >{{'cmx-breadcrumbs.' +breadcrumb.title | translate}}</a></h3>
    } @else {
      <h2>{{'cmx-breadcrumbs.' + breadcrumb.title | translate}}</h2>
    }
  }
