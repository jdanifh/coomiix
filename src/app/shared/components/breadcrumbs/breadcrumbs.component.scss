:host {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  align-items: center;
  margin-left: 100px;

  @media (width < 600px) {
    order: 5;
    flex: 1 1 100%;
    margin-left: 50px;
  }
}

a {
  text-decoration: none;
  color: var(--primary-text-color);
}

a + a, a + h2, h3 + h3, h3 + h2 {
  &::before {
    content: "â¬¤";
    padding: 5px;
    font-size: .25em;
    vertical-align: middle;
    position: relative;
    top: -.2em;
  }
}
