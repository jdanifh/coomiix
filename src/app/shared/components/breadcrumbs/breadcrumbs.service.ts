import { inject, Injectable, signal, WritableSignal } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { ActivatedRouteSnapshot, ActivationEnd, ActivationStart, Router, UrlSegment } from '@angular/router';
import { filter } from 'rxjs';

export interface Breadcrumb {
  title: string,
  path: string,
}

@Injectable({
  providedIn: 'root'
})
export class BreadcrumbsService {

  private _router = inject(Router);

  private _breadcrumbs : WritableSignal<Breadcrumb[]> = signal([]);
  public get breadcrumbs() {
    return this._breadcrumbs.asReadonly();
  }

  constructor() {
    this._router.events.pipe(
      takeUntilDestroyed(),
      filter(event => event instanceof ActivationStart),
    ).subscribe(() => this.reset());

    this._router.events.pipe(
      takeUntilDestroyed(),
      filter(event => event instanceof ActivationEnd),
    ).subscribe(event => this._addRouterBreadcrumb(event));
  }

  private _addRouterBreadcrumb(event: ActivationEnd) {
    if (event.snapshot.title) {
      const title = event.snapshot.title;
      const breadcrumb = {title: title, path: this._getPath(event.snapshot.pathFromRoot)};
      this._breadcrumbs.update(value => [breadcrumb, ...value]);
    }
  }

  private _getPath(route: ActivatedRouteSnapshot[]) {
    return route.reduce((acc, current) => acc + this._getUrl(current.url), '');
  }
  private _getUrl(path: UrlSegment[]) {
    return path.reduce((acc, current) => acc + '/' + current.toString(), '');
  }

  public addBreadcrumb(breadcrumb: Breadcrumb) {
    this._breadcrumbs.update(value => [...value, breadcrumb]);
  }

  public removeBreadcrumb() {
    this._breadcrumbs.update(value => value.slice(0, value.length -1));
  }

  public setBreadcrumbs(breadcrumbs: Breadcrumb[]) {
    this._breadcrumbs.set(breadcrumbs);
  }

  public reset() {
    this._breadcrumbs.set([]);
  }
}
