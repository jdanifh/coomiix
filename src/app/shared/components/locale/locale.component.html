<button [matMenuTriggerFor]="menu" mat-icon-button aria-label="change locale">
  <mat-icon aria-hidden="true">language</mat-icon>
</button>
<mat-menu #menu="matMenu">
  @for (locale of locales; track locale) {
    <button (click)="changeLocale(locale)" mat-menu-item>
      <span class="locale">{{ locale | uppercase | slice:0:2 }}</span>
      <span>{{ locale }}</span>
    </button>
  }
</mat-menu>
