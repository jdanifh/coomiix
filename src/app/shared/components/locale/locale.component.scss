:host {
  margin-right: 25px;
  margin-left: auto;
}

.locale {
  font-weight: 700;
  margin-right: 25px;
}
