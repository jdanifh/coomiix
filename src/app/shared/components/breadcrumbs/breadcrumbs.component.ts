import { ChangeDetectionStrategy, Component, inject } from '@angular/core';
import { RouterModule } from '@angular/router';
import { TranslateModule } from '@ngx-translate/core';
import { BreadcrumbsService } from './breadcrumbs.service';

@Component({
  selector: 'cmx-breadcrumbs',
  imports: [
    RouterModule,
    TranslateModule
  ],
  templateUrl: './breadcrumbs.component.html',
  styleUrl: './breadcrumbs.component.scss',
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class BreadcrumbsComponent {

  private _service = inject(BreadcrumbsService);
  public breadcrumbs = this._service.breadcrumbs;

}
