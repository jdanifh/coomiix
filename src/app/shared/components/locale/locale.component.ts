import { ChangeDetectionStrategy, Component, inject } from '@angular/core';
import { MatButtonModule } from '@angular/material/button';
import { MatIconModule } from '@angular/material/icon';
import { MatMenuModule } from '@angular/material/menu';
import { LocaleService } from './locale.service';
import { CommonModule } from '@angular/common';

@Component({
  selector: 'cmx-locale',
  imports: [
    CommonModule,
    MatIconModule,
    MatButtonModule,
    MatMenuModule,
  ],
  templateUrl: './locale.component.html',
  styleUrl: './locale.component.scss',
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class LocaleComponent {

  private _service = inject(LocaleService);

  public locale = this._service.locale;
  public locales = this._service.locales;

  public changeLocale(locale: string) {
    this._service.changeLocale(locale);
  }

}
