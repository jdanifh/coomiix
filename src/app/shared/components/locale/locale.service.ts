import { inject, Injectable, signal } from '@angular/core';
import { DateAdapter } from '@angular/material/core';
import { TranslateService } from '@ngx-translate/core';
import { DEFAULT_LOCALE, LOCALES } from '../../../core/providers/locale/locale.provider';

export const LOCALE_KEY = 'LOCALE_ID';

@Injectable({
  providedIn: 'root'
})
export class LocaleService {

  private _locale = signal<string>(DEFAULT_LOCALE);
  private _adapter = inject(DateAdapter);
  private _service = inject(TranslateService);

  constructor() {
    this._service.addLangs(Object.values(LOCALES));
    const savedLocale = localStorage.getItem(LOCALE_KEY);
    if (savedLocale) {
      this.changeLocale(savedLocale);
    } else {
      this.changeLocale(DEFAULT_LOCALE);
    }
  }

  public get locale() {
    return this._locale.asReadonly();
  }

  public get locales() {
    return this._service.getLangs();
  }

  public changeLocale(locale: string) {
    this._locale.set(locale);
    this._service.use(locale);
    this._adapter.setLocale(locale);
    localStorage.setItem(LOCALE_KEY, locale);
  }
}
