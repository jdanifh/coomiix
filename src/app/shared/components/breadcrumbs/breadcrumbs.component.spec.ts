import { ComponentFixture, TestBed } from '@angular/core/testing';

import { BreadcrumbsComponent } from './breadcrumbs.component';
import { By } from '@angular/platform-browser';
import { BreadcrumbsService } from './breadcrumbs.service';
import { provideRouter } from '@angular/router';
import { routes } from '../../../app.routes';
import { provideLocalesTest } from '../../../core/providers/locale/locale.provider';

describe('BreadcrumbsComponent', () => {
  let component: BreadcrumbsComponent;
  let fixture: ComponentFixture<BreadcrumbsComponent>;
  let service: BreadcrumbsService;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [BreadcrumbsComponent],
      providers: [
        provideRouter(routes),
        provideLocalesTest()
      ]
    })
    .compileComponents();

    fixture = TestBed.createComponent(BreadcrumbsComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    service = TestBed.inject(BreadcrumbsService);
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should render breadcrumbs', () => {
    const breadcrumb = {title: 'Home', path: ''};
    service.addBreadcrumb(breadcrumb);
    fixture.detectChanges();
    let last = fixture.debugElement.query(By.css('h2'));
    expect(last.nativeElement.textContent).toContain(breadcrumb.title);

    const breadcrumb2 = {title: 'Dashboard', path: '/dashboard'};
    service.addBreadcrumb(breadcrumb2);
    fixture.detectChanges();
    last = fixture.debugElement.query(By.css('h2'));
    expect(last.nativeElement.textContent).toContain(breadcrumb2.title);
    const notLast = fixture.debugElement.query(By.css('h3'));
    expect(notLast.nativeElement.textContent).toContain(breadcrumb.title);
  })

  it(`should match snapshot`, () => {
    expect(fixture.nativeElement.innerHTML).toMatchSnapshot();
  })

});
