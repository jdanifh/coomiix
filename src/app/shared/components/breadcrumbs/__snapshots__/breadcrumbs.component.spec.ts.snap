// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BreadcrumbsComponent should match snapshot 1`] = `"<!--container-->"`;
